lib/bolt/outputter.rb

```diff
@@ -21,6 +21,76 @@ def initialize(color, verbose, trace, stream = $stdout)
       @trace = trace
       @stream = stream
     end
+
+    def indent(indent, string)
+      indent = ' ' * indent
+      string.gsub(/^/, indent.to_s)
+    end
+
+    def print_message_event(event)
+      print_message(stringify(event[:message]))
+    end
+
+    def print_message
+      raise NotImplementedError, "print_message() must be implemented by the outputter class"
+    end
+
+    def stringify(message)
+      formatted = format_message(message)
+      if formatted.is_a?(Hash) || formatted.is_a?(Array)
+        ::JSON.pretty_generate(formatted)
+      else
+        formatted
+      end
+    end
+
+    def format_message(message)
+      case message
+      when Array
+        message.map { |item| format_message(item) }
+      when Bolt::ApplyResult
+        format_apply_result(message)
+      when Bolt::Result, Bolt::ResultSet
+        # This is equivalent to to_s, but formattable
+        message.to_data
+      when Bolt::RunFailure
+        formatted_resultset = message.result_set.to_data
+        message.to_h.merge('result_set' => formatted_resultset)
+      when Bolt::PlanResult
```

Comment:
I don't think we need this case. A `Bolt::PlanResult` instance is only constructed within Bolt when the outermost plan run finishes, so it shouldn't be available to `out::message` under normal circumstances. Technically a Ruby function could construct one on its own, but that's a corner case and probably not worth having special handling for.

Reply:
You're totally right - I thought the result of a `run_plan` would be this type, but it's just a ResultSet. Removed!